### PR TITLE
fix: Handle preproc statistics when no tuples are found

### PR DESCRIPTION
**Pull request by stygerma**
- If no image tuples are found for a bag return `None` for the overlap statistics  
- Make the `data/proc` directory in io class

### CODE REVIEW

File: sips/preprocessing/io.py [hunk 1]
```diff
@@ -38,6 +38,7 @@ def find_config_id(config: DatasetsConfig) -> int:
         Configuration ID.
 
     """
+    Path(config.proc_path).mkdir(parents=True, exist_ok=True)
```
Comment (tstreule) [inline sips/preprocessing/io.py:41]: I don't see why we should create the directory here.

File: sips/preprocessing/processor.py [hunk 2]
```diff
@@ -400,14 +406,12 @@ def get_summary_statistics(self, rosbags: Iterable[str]) -> dict[str, float]:
...
         stats = self.io.read_statistics(rosbags)
-        # Get keys from arbitrary rosbag
-        keys = stats[next(iter(stats))].keys()
 
         # Make summary statistics
         summary: dict[str, float] = defaultdict(float)
         for bag in stats:
             bag_stats = stats[bag]
-            for key in keys:
+            for key in stats[bag].keys():
```
Comment (tstreule) [inline sips/preprocessing/processor.py:414]: Use the `bag_stats` variable

File: sips/preprocessing/processor.py [hunk 3]
```diff
@@ -198,7 +198,10 @@ def _process(
         )
 
         # Gather statistics
-        statistics: dict[str, float] = {**stat_read, **stat_overlap}
+        if stat_overlap:
+            statistics: dict[str, float] = {**stat_read, **stat_overlap}
+        else:
+            statistics = stat_read
```
Comment (tstreule) [inline sips/preprocessing/processor.py:204]: Undo (see review comment in `find_overlapping_tuples`)

File: sips/preprocessing/processor.py [hunk 4]
```diff
@@ -330,7 +333,7 @@ def interpolate_pose(
 
     def find_overlapping_tuples(
         self, image_messages: ImageDict, pose_data: PoseDict
-    ) -> tuple[StatisticsDict, ImageTuples]:
+    ) -> tuple[StatisticsDict | None, ImageTuples]:
```
Comment (tstreule) [inline sips/preprocessing/processor.py:336]: Undo (see review comment in `find_overlapping_tuples`)

File: sips/preprocessing/processor.py [hunk 5]
```diff
@@ -376,6 +379,9 @@ def find_overlapping_tuples(
 
         if overlapping_tuples:  # prevent division by zero
             stats["overlap_ratio_kept_mean"] /= len(overlapping_tuples)
+        else:
+            warn(f"No overlapping tuples found")
+            return None, overlapping_tuples
```
Comment (tstreule) [inline sips/preprocessing/processor.py:384]: Good one!

Raise an exception instead. Either a `ReaderError` (makes few sense but simple) or use another exception and add that exception to the decorator in the `_process` method

Comment (tstreule) [reply sips/preprocessing/processor.py:384]: When we have no tuples, it doesn't make sense to store anything. Hence, raising an Exception will result in skipping the bag and logging a (warning) message in the console.

**Review by tstreule — CHANGES_REQUESTED** [review]

File: sips/preprocessing/processor.py (hunk 2 above)
Comment (stygerma) [reply sips/preprocessing/processor.py]: done

File: sips/preprocessing/processor.py (hunk 3 above)
Comment (stygerma) [reply sips/preprocessing/processor.py]: done

File: sips/preprocessing/processor.py (hunk 4 above)
Comment (stygerma) [reply sips/preprocessing/processor.py]: done

File: sips/preprocessing/processor.py (hunk 5 above)
Comment (stygerma) [reply sips/preprocessing/processor.py]: done

File: sips/preprocessing/io.py (hunk 1 above)
Comment (tstreule) [reply sips/preprocessing/io.py:41]: True. I propose to write it a bit further below:

```python
# Handle when no match is found
if config_id is None:
    config_id = ...
    info[...] = ...
    # Write file
    config_path.parent.mkdir(parents=True, exist_ok=True)  # Here
    with open(...) as f:
        ...
```

Comment (stygerma) [reply sips/preprocessing/io.py]: done

**Review by tstreule — APPROVED** [review]

**Pull request merged**